identity/identity.go

```diff
@@ -6,49 +6,110 @@ import (
+	return handler(doAuth, rchttp.DefaultClient)
+}
+
+type identityResponse struct {
+	Identifier string `json:"identifier"`
+}
+
+func handler(doAuth bool, cli HttpClient) func(http.Handler) http.Handler {
 	return func(h http.Handler) http.Handler {
 		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-			if doAuth && len(req.Header.Get("X-Florence-Token")) > 0 {
 
-				// set a default zebedee value if it isn't set
-				if len(zebedeeURL) == 0 {
-					zebedeeURL = "http://localhost:8082"
-				}
+			ctx := req.Context()
 
-				cli := rchttp.DefaultClient
+			if doAuth {
+				florenceToken := req.Header.Get(florenceHeaderKey)
+				authToken := req.Header.Get(authHeaderKey)
 
-				zebReq, err := http.NewRequest("GET", zebedeeURL+"/permission", nil)
-				if err != nil {
-					log.ErrorR(req, err, nil)
-					w.WriteHeader(http.StatusInternalServerError)
-					return
-				}
+				isUserReq := len(florenceToken) > 0
+				isServiceReq := len(authToken) > 0
 
-				zebReq.Header.Set("X-Florence-Token", req.Header.Get("X-Florence-Token"))
+				if isUserReq || isServiceReq {
 
-				resp, err := cli.Do(req.Context(), zebReq)
-				if err != nil {
-					log.ErrorR(req, err, nil)
-					w.WriteHeader(http.StatusInternalServerError)
-					return
-				}
+					// set a default zebedee value if it isn't set
+					if len(zebedeeURL) == 0 {
+						zebedeeURL = "http://localhost:8082"
+					}
 
-				// Check to see if the florence user is authorised
-				if resp.StatusCode != http.StatusOK {
-					w.WriteHeader(resp.StatusCode)
-					return
-				}
+					zebReq, err := http.NewRequest("GET", zebedeeURL+"/identity", nil)
+					if err != nil {
+						log.ErrorR(req, err, nil)
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+
+					if isUserReq {
+						zebReq.Header.Set(florenceHeaderKey, florenceToken)
+					} else if isServiceReq {
+						zebReq.Header.Set(authHeaderKey, authToken)
+					}
 
+					resp, err := cli.Do(ctx, zebReq)
+					if err != nil {
+						log.ErrorR(req, err, nil)
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+
+					// Check to see if the florence user is authorised
+					if resp.StatusCode != http.StatusOK {
+						w.WriteHeader(resp.StatusCode)
+						return
+					}
+
+					identityResp, err := unmarshalResponse(resp)
+					if err != nil {
+						log.ErrorR(req, err, nil)
+						w.WriteHeader(http.StatusInternalServerError)
+						return
+					}
+
+					userIdentity := identityResp.Identifier
+					if !isUserReq {
+						userIdentity = req.Header.Get(userIdentityKey)
+					}
+
+					ctx = context.WithValue(ctx, userIdentityKey, userIdentity)
+					ctx = context.WithValue(ctx, callerIdentityKey, identityResp.Identifier)
+				}
 			}
 
-			h.ServeHTTP(w, req)
+			h.ServeHTTP(w, req.WithContext(ctx))
 		})
 	}
 }
+
+func unmarshalResponse(resp *http.Response) (*identityResponse, error) {
+
+	var identityResp *identityResponse
```

Comment:
consider naming the variables in the response part of function signature, e.g:

```go
func unmarshalResponse(resp *http.Response) (identityResp *identityResponse, err error) {
```

Reply:
Done 👍